shiki/WordPress-Android
# Import from Giphy in Media Library

![dec-01-2018 17-17-03](https://user-images.githubusercontent.com/198826/49334196-fe91e180-f58c-11e8-9a7a-21db4ac29a8b.gif)

This is a part of the Giphy picker feature (wordpress-mobile/WordPress-Android#8627). It adds importing of GIFs from Giphy from the Media Library.

## Workflow

This is an overview of what happens when the user presses the "Add" button in the Giphy picker:

1. `GiphyPickerViewModel.state` is changed to `State.DOWNLOADING` which disables the _Preview_, _Add_, and _Search_ controls. _Selecting_ is also disabled.
2. The _Add_ button is replaced with an indeterminate progress bar.
3. The selected GIFs are downloaded concurrently. For each of them, a `MediaModel` is created and saved using `MediaActionBuilder.newUpdateMediaAction()`. This all happens in `GiphyMediaFetcher`.
4. When the download finishes, the `ViewModel` posts a `DownloadResult` event which contains the saved `MediaModel` objects.
5. The `Activity` finishes and returns the _local ids_ of those `MediaModel` objects.
6. `MediaBrowserActivity` receives the local ids, loads their corresponding `MediaModel` objects using `MediaStore.getMediaWithLocalId` and adds them to the `UploadService` for uploading.

## Progress Bar

I started to implement the progress bar UI using `ProgressDialog` but then found out that that is now deprecated. I opted to use a circular progress bar which replaces the _Add_ button instead. I based this from the [Integrating with actions](https://material.io/design/components/progress-indicators.html#circular-progress-indicators) section in Material Design.

Please let me know if this does not coincide with the WordPress app design conventions.

## Reused Layout

Previously, we made `media_picker_activity` reusable. This change adds a `ProgressBar` to that layout which makes this layout not too generic anymore. I also just realized that that layout has an existing `ProgressBar` used for endless scrolling that the Giphy Picker doesn't use. These little differences seem okay to me but I'm not sure up to when we should tolerate this.

## Future Improvements

- I've been wondering whether we should add a limit to the number of GIFs that users can select. It's probably unlikely that users would select too many which will cause out-of-memory errors. This can be a future improvement.
- When the _Add_ and _Preview_ buttons are disabled, their colors don't change. I'm not sure what colors to change them to.
- A determinate progress bar may be helpful. We could probably do some user research first whether this is really needed.

## Testing

1. Create a [Giphy API key](https://developers.giphy.com/docs/) and paste it in your `gradle.properties` as the value for `wp.giphy.api_key`
2. Go to the Media Library
3. Click on the _Add_ button at the top right. Select the _Choose from Giphy_ option
4. In the Giphy picker, search and pick some GIFs
5. Press the _Add_ button. The selected GIFs should be downloaded and the picker should be closed. The upload will commence in the Media Library.

Some methods used by other parts of the app were changed. These may have regressions:

- Places that use Photo Picker like _Me_ or the _Media Library_
- Stock Media Picker
